pkg/daemon/daemon.go

```diff
@@ -359,54 +380,65 @@ func (dn *Daemon) getStateAndConfigs() (*stateAndConfigs, error) {
-// completed a config, and are coming up after a reboot.
-func (dn *Daemon) isConfigCompletedPending(pending string) (bool, error) {
+// getPendingConfig loads the JSON state we cache across attempting to apply
+// a config+reboot.  If no pending state is available, ("", nil) will be returned.
+// The bootID is stored in the pending state; if it is unchanged, we assume
+// that we failed to reboot; that for now should be a fatal error, in order to avoid
+// reboot loops.
+func (dn *Daemon) getPendingConfig() (string, error) {
 	s, err := ioutil.ReadFile(pathStateJSON)
 	if err != nil {
 		if !os.IsNotExist(err) {
-			return false, errors.Wrapf(err, "Loading transient state")
+			return "", errors.Wrapf(err, "Loading transient state")
 		} else {
-			return false, nil
+			return "", nil
 		}
 	}
-	var t transientState
-	if err := json.Unmarshal([]byte(s), &t); err != nil {
-		return false, errors.Wrapf(err, "Parsing transient state")
+	var p pendingConfigState
+	if err := json.Unmarshal([]byte(s), &p); err != nil {
+		return "", errors.Wrapf(err, "Parsing transient state")
 	}
 
-	if t.PendingConfig != pending {
-		glog.Infof("Found pending config %s, expected %s", t.PendingConfig, pending)
-		return false, nil
-	}
-	if t.BootID == dn.bootID {
-		dn.logSystem(fmt.Sprintf("pending bootID %s matches current! Failed to reboot?", dn.bootID))
-		return false, nil
+	if p.BootID == dn.bootID {
+		return "", fmt.Errorf("pending config %s bootID %s matches current! Failed to reboot?", p.PendingConfig, dn.bootID)
 	}
-	return true, nil
+	return p.PendingConfig, nil
 }
 
-// CheckStateOnBoot is responsible for checking whether the node has
-// degraded, and if not, whether an update is required immediately.
-// The flow goes something like this -
-// 1. Sanity check if we're in a degraded state. If yes, handle appropriately.
-// 2. we restarted for some reason. the happy path reason we restarted is
-//    because of a machine reboot. validate the current machine state is the
-//    desired machine state. if we aren't try updating again. if we are, update
-//    the current state annotation accordingly.
+// CheckStateOnBoot is a core entrypoint for our state machine.
+// It determines whether we're in our desired state, or if we're
+// transitioning between states, and whether or not we need to update
+// to a new state.
+//
+// Some more background in this PR: https://github.com/openshift/machine-config-operator/pull/245
 func (dn *Daemon) CheckStateOnBoot() error {
-	stateAndConfigs, err := dn.getStateAndConfigs()
+	pendingConfigName, err := dn.getPendingConfig()
 	if err != nil {
-		// TODO; don't go into degraded here
-		dn.nodeWriter.SetUpdateDegradedIgnoreErr(err, dn.kubeClient.CoreV1().Nodes(), dn.name)
-		stateAndConfigs.state = MachineConfigDaemonStateDegraded
+		return err
+	}
+	stateAndConfigs, err := dn.getStateAndConfigs(pendingConfigName)
+	if err != nil {
+		return err
 	}
 	if stateAndConfigs.state == MachineConfigDaemonStateDegraded {
```

Comment:
Let's change the `select {}` here to just return an error so we sleep in the new `EnterDegradedState()`?

Reply:
I started to make this change then remembered there's a reason I wrote it this way; it's that we want to clearly distinguish between the case of "we're explicitly transitioning the system to degraded because $REASON" from "we started up degraded".

The idea here of course is that `EnterDegradedState` is where we'd implement https://github.com/openshift/machine-config-operator/issues/234